File: C++/LinkedList_Heap_PriorityQueue_Hard_MergeKSortedLists_23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) 
    {
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> pq;

        if(lists.empty())
        {
            return NULL;
        }

        for(ListNode* head : lists)
        {
            if(head != NULL)
            {
                pq.push({head -> val, head});
            }
        }

        ListNode* dummy = new ListNode();
        ListNode* ptr = dummy;

        while(!pq.empty())
        {
            ListNode* node = pq.top().second;
            dummy -> next = node;
            dummy = node;
            pq.pop();
            if(node -> next != NULL)
            {
                pq.push({node -> next -> val, node -> next});
            }
        }

        if(ptr -> next != NULL)
        {
            return ptr -> next;
        }
        else
        {
            return NULL;
        }

    }
};

File: Python/LinkedList_Heap_PriorityQueue_Hard_MergeKSortedLists_23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []

        if not lists: 
            return None

        z = len(lists)

        for i, head in enumerate(lists):
            if head:
                heapq.heappush(heap, (head.val, i, head))

        dummy = ListNode(0, None)
        ptr = dummy

        while heap:
            node = heapq.heappop(heap)[2]
            dummy.next = node
            dummy = node

            if node.next:
                heapq.heappush(heap, (node.next.val, z, node.next))
                z += 1

        return ptr.next

        